File: packages/semantics/src/default-schema.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Default schema code derived from a definition's kind, applied only when a
 * file has no explicit `schema` directive. An explicit directive always wins
 * for the whole file. Mirrors the namespace fallback (`namespace || def.kind`)
 * but per the normative kind→schema map in
 * `docs/features/pkg-schema-defaults/INDEX.md`.
 *
 * `kind` is the camelCase `def.kind` string (e.g. `er2dbEntity`, `drillMap`).
 * Unknown kinds fall back to `db` — kept identical to the Kotlin twin
 * (`defaultSchemaForKind` in `Kinds.kt`).
 */
/**
 * Conventional default namespace for a schema, applied when a file declares a
 * `schema` but no explicit `namespace` (and, symmetrically, to the schema
 * derived for schema-less files). Mirrors the architecture's manifest default
 * `namespaces = { db = "dbo", … }` (architecture.md §5).
 *
 * Scoped to `db` (the SQL-default `dbo` schema): without it, a `schema db` file
 * with no namespace registers tables/columns under per-kind segments
 * (`db.table.<t>`, `db.table.<t>.<col>`), so the canonical fully-qualified
 * `db.dbo.<table>.<column>` references fail to resolve. Every other schema
 * returns '' so the caller keeps the per-kind fallback (`def.kind`): `er`
 * already coincides with its `entity` namespace, `binding` addresses symbols per
 * kind (`binding.er2dbEntity.…`, relied on by the inline-mapping synthesizer), and
 * `cnc`/`query` files carry explicit namespaces in practice.
 */
import type { ModelCode } from './qname.js';

export function defaultNamespaceForSchema(schemaCode: string): string {
  return schemaCode === 'db' ? 'dbo' : '';
}

/**
 * The single-valued kind→model map (D4/D14/D15) and the one source of truth for
 * model derivation. `query`/`drillMap` → `db` (D14 — there is no `query` model);
 * `role`/`er2cncRole` → `cnc` (D15, schema-less); er2db / md2db binding kinds →
 * `binding`; MD logical kinds → `md`; everything else (table, view, …) → `db`.
 *
 * `defaultSchemaForKind` is a deprecated alias kept for the existing call sites
 * (`reference-index`, `mapping-references`, lint rules) — it returns the same
 * value. `qname.ts` re-exports `modelForKind` so the public API name is stable.
 * Mirrored byte-for-byte in Kotlin (`Kinds.kt`) and Python (`qname.py`).
 */
export function modelForKind(kind: string): ModelCode {
  switch (kind) {
    case 'entity':
    case 'attribute':
    case 'relation':
      return 'er';
    case 'er2dbEntity':
    case 'er2dbAttribute':
    case 'er2dbRelation':
    case 'md2dbCubelet':
    case 'md2dbDomain':
    case 'md2dbMap':
    case 'md2erCubelet':
      return 'binding';
    case 'role':
    case 'er2cncRole':
      return 'cnc';
    // MD logical kinds → `md`.
    case 'mdDomain':
    case 'dimension':
    case 'mdMap':
    case 'hierarchy':
    case 'measure':
    case 'cubelet':
      return 'md';
    // v4.1 — world model kinds (ttr-metadata M0).
    case 'world':
    case 'engine':
    case 'executor':
    case 'storage':
    case 'worldSchema':
      return 'world';
    // D14 — query + drillMap are db-layer objects (no separate `query` model).
    case 'query':
    case 'drillMap':
    case 'project':
    case 'table':
    case 'view':
    case 'column':
    case 'index':
    case 'constraint':
    case 'fk':
    case 'procedure':
      return 'db';
    default:
      return 'db';
  }
}

/** @deprecated Use {@link modelForKind}. Kept as an alias for existing callers. */
export const defaultSchemaForKind: (kind: string) => ModelCode = modelForKind;

/**
 * The symbol-table namespace segment for a def kind, where it differs from the
 * camelCase `def.kind`. MD logical/binding kinds map to the contracts §5
 * namespaces (`md.domain.*`, `md.map.*`, `binding.md2db_cubelet.*`, …) rather
 * than the raw kind (`mdDomain`/`mdMap`). Returns `''` for every other kind so
 * the caller keeps the existing `def.kind` fallback.
 */
export function namespaceForKind(kind: string): string {
  switch (kind) {
    case 'mdDomain':
      return 'domain';
    case 'mdMap':
      return 'map';
    case 'dimension':
      return 'dimension';
    case 'hierarchy':
      return 'hierarchy';
    case 'measure':
      return 'measure';
    case 'cubelet':
      return 'cubelet';
    case 'md2dbCubelet':
      return 'md2db_cubelet';
    case 'md2dbDomain':
      return 'md2db_domain';
    case 'md2dbMap':
      return 'md2db_map';
    case 'md2erCubelet':
      return 'md2er_cubelet';
    default:
      return '';
  }
}

File: packages/python/ttr-parser/src/ttr_parser/semantics/default_schema.py
"""Default schema/namespace derivation (← `default-schema.ts`).

Applied only when a file has no explicit `schema` directive (kind → schema) or
no explicit `namespace` (schema → namespace). An explicit directive always wins.
Kept identical to the TS twin so the symbol table builds byte-identical qnames
(pinned by the §5.1 conformance dump).

`kind` here is the **Python** `Definition.kind` string (snake_case, e.g.
`er2db_entity`) — see `default_schema_for_kind`. The qname *namespace-fallback*
segment, however, must match the TS camelCase form; that snake→camel mapping
lives in `symbol_table._kind_segment`, not here.
"""

from __future__ import annotations


def default_namespace_for_schema(schema_code: str) -> str:
    """`db` → `dbo`; every other schema → '' (caller keeps the per-kind fallback)."""
    return "dbo" if schema_code == "db" else ""


def default_schema_for_kind(kind: str) -> str:
    """Deprecated alias of :func:`model_for_kind` — the single kind→model source of
    truth (D4/D14/D15). Applied only when a file has no explicit ``schema``
    directive. ``query``/``drill_map`` → ``db`` (D14, not the retired ``query``
    value). Mirrors the TS twin, where ``defaultSchemaForKind`` aliases
    ``modelForKind``."""
    return model_for_kind(kind)


# The TS/Kotlin `def.kind` is camelCase (`er2dbEntity`); the Python model uses
# snake_case (`er2db_entity`). Qnames and the enclosing-qname namespace-fallback
# segment must use the camelCase form so they are byte-identical to the §5.1
# golden (e.g. `binding.er2dbEntity.x`, and `kindOf` membership in `references`).
_KIND_SEGMENT: dict[str, str] = {
    "er2db_entity": "er2dbEntity",
    "er2db_attribute": "er2dbAttribute",
    "er2db_relation": "er2dbRelation",
    "er2cnc_role": "er2cncRole",
    "drill_map": "drillMap",
}


def kind_segment(kind: str) -> str:
    """The camelCase (TS) form of a Python snake_case `Definition.kind`."""
    return _KIND_SEGMENT.get(kind, kind)


# ---------------------------------------------------------------------------
# v4.0 uniform key (← `qname.ts` modelForKind / namespaceForKind / buildCanonicalKey)
# ---------------------------------------------------------------------------

MODEL_CODES: frozenset[str] = frozenset({"db", "er", "md", "binding", "cnc", "world"})

_MODEL_BY_KIND: dict[str, str] = {
    "entity": "er",
    "attribute": "er",
    "relation": "er",
    "er2db_entity": "binding",
    "er2db_attribute": "binding",
    "er2db_relation": "binding",
    "md2db_cubelet": "binding",
    "md2db_domain": "binding",
    "md2db_map": "binding",
    "md2er_cubelet": "binding",
    "role": "cnc",
    "er2cnc_role": "cnc",
    "md_domain": "md",
    "dimension": "md",
    "md_map": "md",
    "hierarchy": "md",
    "measure": "md",
    "cubelet": "md",
    "world": "world",
    "engine": "world",
    "executor": "world",
    "storage": "world",
    "worldSchema": "world",
}


def model_for_kind(kind: str) -> str:
    """Kind → model layer (D14/D15). `query`/`drill_map` and everything not
    explicitly mapped → `db`. Mirrors TS `modelForKind`."""
    return _MODEL_BY_KIND.get(kind, "db")


_NAMESPACE_BY_KIND: dict[str, str] = {
    "md_domain": "domain",
    "md_map": "map",
    "dimension": "dimension",
    "hierarchy": "hierarchy",
    "measure": "measure",
    "cubelet": "cubelet",
    "md2db_cubelet": "md2db_cubelet",
    "md2db_domain": "md2db_domain",
    "md2db_map": "md2db_map",
    "md2er_cubelet": "md2er_cubelet",
}


def namespace_for_kind(kind: str) -> str:
    """The kind-segment namespace alias (MD kinds), else '' (caller keeps the
    camelCase kind segment). Mirrors TS `namespaceForKind`."""
    return _NAMESPACE_BY_KIND.get(kind, "")


def build_canonical_key(
    package_name: str, schema_id: str, kind: str, parts: list[str]
) -> str:
    """The single v4.0 uniform-key builder
    `<package>.<model>.<schema?>.<kind>.<parts>` (← TS `buildCanonicalKey`):
    model from the kind, schema slot db-only (default `dbo`), kind segment via
    the namespace alias else the camelCase kind."""
    model = model_for_kind(kind)
    segments: list[str] = []
    if package_name:
        segments.append(package_name)
    segments.append(model)
    if model == "db":
        segments.append(schema_id or "dbo")
    segments.append(namespace_for_kind(kind) or kind_segment(kind))
    segments.extend(parts)
    return ".".join(segments)
